File: app/src/main/java/com/example/carlauncher/someip/NetworkStateTracker.java
package com.example.carlauncher.someip;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/** Android-free reduction of address, route and per-app blocking observations. */
final class NetworkStateTracker {
    static final class State {
        final String iface;
        final boolean available;
        final boolean route;
        State(String iface, boolean route) {
            this.iface = iface;
            this.available = !iface.isEmpty();
            this.route = available && route;
        }
    }
    private final String address;
    private final boolean requireBlockedVerdict;
    private final Map<Object, State> links = new HashMap<>();
    private final Map<Object, Boolean> blocked = new HashMap<>();

    NetworkStateTracker(String address, boolean requireBlockedVerdict) {
        this.address = address;
        this.requireBlockedVerdict = requireBlockedVerdict;
    }
    void link(Object network, String iface, String[] addresses, boolean route) {
        links.put(network, new State(iface != null && Arrays.asList(addresses).contains(address) ? iface : "", route));
    }
    void blocked(Object network, boolean value) { blocked.put(network, value); }
    void lost(Object network) { links.remove(network); blocked.remove(network); }
    void clear() { links.clear(); blocked.clear(); }
    State snapshot() {
        State selected = new State("", false);
        for (Map.Entry<Object, State> entry : links.entrySet()) {
            Boolean denied = blocked.get(entry.getKey());
            if (Boolean.TRUE.equals(denied) || (requireBlockedVerdict && denied == null)) continue;
            State candidate = entry.getValue();
            if (!candidate.available) continue;
            if (!selected.available || (candidate.route && !selected.route)
                    || (candidate.route == selected.route && candidate.iface.compareTo(selected.iface) < 0)) {
                selected = candidate;
            }
        }
        return selected;
    }
}

File: app/src/main/java/com/example/carlauncher/someip/NetworkStateTracker.kt
package com.example.carlauncher.someip

import java.util.Arrays
import java.util.HashMap

/* Android-free reduction of address, route and per-app blocking observations. */
internal class NetworkStateTracker(
    private val address: String?,
    private val requireBlockedVerdict: Boolean,
) {
    private val links: MutableMap<Any?, State?> = HashMap<Any?, State?>()
    private val blocked: MutableMap<Any?, Boolean?> = HashMap<Any?, Boolean?>()

    internal class State(val iface: String, route: Boolean) {
        val available: Boolean
        val route: Boolean

        init {
            this.available = !iface!!.isEmpty()
            this.route = available && route
        }
    }

    fun link(network: Any?, iface: String?, addresses: Array<out String?>, route: Boolean) {
        links.put(
            network,
            State(
                if (iface != null && Arrays.asList<String?>(*addresses)!!.contains(address)) iface
                else "",
                route,
            ),
        )
    }

    fun blocked(network: Any?, value: Boolean) {
        blocked.put(network, value)
    }

    fun lost(network: Any?) {
        links.remove(network)
        blocked.remove(network)
    }

    fun clear() {
        links.clear()
        blocked.clear()
    }

    fun snapshot(): State {
        var selected: State = State("", false)
        for (entry in links.entries) {
            val denied = blocked.get(entry!!.key)
            if (java.lang.Boolean.TRUE == denied || (requireBlockedVerdict && denied == null))
                continue
            val candidate = entry!!.value
            if (!candidate!!.available) continue
            if (
                (!selected!!.available ||
                    (candidate!!.route && !selected!!.route) ||
                    (candidate!!.route == selected!!.route &&
                        candidate!!.iface!!.compareTo(selected!!.iface) < 0))
            ) {
                selected = candidate
            }
        }
        return selected
    }
}
